src/Microsoft.DotNet.Build.Tasks/Targets/publishtest.targets

```diff
@@ -38,6 +52,40 @@
       <Output TaskParameter="ResolvedCopyLocal" ItemName="TestCopyLocal" />
     </ResolveNuGetPackages>
 
+    <ResolveNuGetPackageAssets AdditonalProjectJsonFiles="@(ProjectJsons)"
+                               Architecture="$(PlatformTarget)"
+                               Configuration="$(NuGetConfiguration)"
+                               IngoreLockFile="$(NuGetIngoreLockFile)"
+                               Language="$(Language)"
+                               PackageRoot="$(PackagesDir)"
+                               ProjectFile="$(MSBuildProjectFullPath)"
+                               TargetFrameworkMonikers="@(TestTargetFramework)"
+                               TargetPlatformMonikers="$(TargetPlatformMoniker)"
+                               TransitiveProjectReferences="">
+
+      <Output TaskParameter="ResolvedCopyLocalItems" ItemName="TestCopyLocal" />
+    </ResolveNuGetPackageAssets>
+
+    <!-- We may have an indirect package reference that we want to replace with a project reference -->
+    <ItemGroup>
+      <!-- Convert to filenames so that we can intersect -->
+      <_ProjectReferenceFilenames Include="@(_ResolvedProjectReferencePaths->'%(FileName)%(Extension)')">
+        <OriginalIdentity>%(Identity)</OriginalIdentity>
+      </_ProjectReferenceFilenames>
+
+      <_TestCopyLocalFileNames Include="@(Reference->'%(FileName)%(Extension)')">
+        <OriginalIdentity>%(Identity)</OriginalIdentity>
+      </_TestCopyLocalFileNames>
+
+      <!-- Intersect project-refs with package-refs -->
+      <_TestCopyLocalFileNamesToRemove Include="@(_ReferencesFileNames->'%(OriginalIdentity)')" Condition="'@(_ProjectReferenceFilenames)' == '@(_ReferencesFileNames)' and '%(Identity)' != ''"/>
+
+      <TestCopyLocal Remove="@(_TestCopyLocalFileNamesToRemove)" />
```

Comment:
I'd prefer to not take the chance/rely on ordering of the copy item group.  I had to filter for refs because CSC complains, so I applied the pattern to everything.

I actually plan on killing most of this target once I can move to a meta-package for tests, so my care-factor is low.